tools/wptrunner/wptrunner/browsers/base.py

```diff
@@ -98,6 +98,10 @@ def __enter__(self):
         self.setup()
         return self
 
+    def update_device_serial_for_executor(self, kwargs):
```

Comment:
Putting this on the generic `Browser` seems unfortunate. Is there no opportunity to update the device serial in `ChromeAndroidBrowserBase`'s `__init__` or `setup` method?

Reply:
This function is to set ["capabilities"]["goog:chromeOptions"]["androidDeviceSerial"] for the executor, so maybe does not belong to __init__ or setup?

Reply:
Instead of doing this you should use the `ExecutorBrowser` class (via the `executor_browser` method) to pass data that's set for a particular `Browser` instance into the corresponding `Executor`.

Reply:
Thank you jpraham for the review. I tried to make change as below, but test shows it did not work.

     def executor_browser(self):
-        return ExecutorBrowser, {"webdriver_url": self.server.url}
+        return ExecutorBrowser, {"webdriver_url": self.server.url,
+                                 "capabilities": {
+                                     "goog:chromeOptions": {
+                                         "androidDeviceSerial": self.device_serial
+                                     }
+                                 }}

Reply:
You will need to add something like 
```py
if hasattr(browser, "capabilities"):
    capabilities.update(browser.capabilities)
```
in `WebDriverProtocol.__init__` in `executorwebdriver.py`. Except in reality you probably need some specific logic to handle update rather than overwrite. You could either write some kind of recursive update method, or just special case the device serial, but that probably requires having a specific subclass of the test type executors for Chrome.

BTW it would also be nice to see this implemented for Firefox, since it should be very similar code.

Reply:
Thanks, I updated per this suggestion and it worked!

For Firefox, I made a change to executormarionette.py. But I did not update firefox_android.py, as I don't know what parameters are required for firefox. Is that ok?